cirq-google/cirq_google/experimental/analog_experiments/generic_analog_circuit.py

```diff
@@ -0,0 +1,154 @@
+    if match is None:
+        raise ValueError(f"Invalid qubit name format: '{qubit_name}'. Expected 'q<row>_<col>'.")
+    return cirq.GridQubit(int(match[1]), int(match[2]))
+
+
+def _coupler_name_from_qubit_pair(qubit_pair: tuple[str, str]) -> str:
+    sorted_pair = sorted(qubit_pair, key=_to_grid_qubit)
+    return f"c_{sorted_pair[0]}_{sorted_pair[1]}"
+
+
+def _get_neighbor_coupler_freqs(
+    qubit_name: str, coupler_g_dict: dict[tuple[str, str], su.ValueOrSymbol]
+) -> dict[str, su.ValueOrSymbol]:
+    """Get neighbor coupler coupling strength g given qubit name."""
+    return {
+        _coupler_name_from_qubit_pair(pair): g
+        for pair, g in coupler_g_dict.items()
+        if qubit_name in pair
+    }
+
+
+class GenericAnalogCircuitBuilder:
+    """Class for making arbitrary analog circuits. The circuit is defined by an
+    AnalogTrajectory object. The class constructs the circuit from AnalogDetune
+    pulses, which automatically calculate the necessary bias amps to both qubits
+    and couplers, using tu.Values from analog calibration whenever available.
+
+    Attributes:
+        trajectory: AnalogTrajectory object defining the circuit
+        g_ramp_shaping: coupling ramps are shaped according to ramp_shape_exp if True
+        qubits: list of qubits in the circuit
+        pairs: list of couplers in the circuit
+        ramp_shape_exp: exponent of g_ramp (g proportional to t^ramp_shape_exp)
+        interpolate_coupling_cal: interpolates between calibrated coupling tu.Values if True
+        linear_qubit_ramp: if True, the qubit ramp is linear. if false, a cosine shaped
+            ramp is used.
+    """
+
+    def __init__(
+        self,
+        trajectory: atu.AnalogTrajectory,
+        g_ramp_shaping: bool = False,
+        ramp_shape_exp: int = 1,
+        interpolate_coupling_cal: bool = False,
+        linear_qubit_ramp: bool = True,
+    ):
+        self.trajectory = trajectory
+        self.g_ramp_shaping = g_ramp_shaping
+        self.ramp_shape_exp = ramp_shape_exp
+        self.interpolate_coupling_cal = interpolate_coupling_cal
+        self.linear_qubit_ramp = linear_qubit_ramp
+
+    def make_circuit(self) -> cirq.Circuit:
+        """Assemble moments described in trajectory."""
+        prev_freq_map = self.trajectory.full_trajectory[0]
+        moments = []
+        for freq_map in self.trajectory.full_trajectory[1:]:
+            if freq_map.is_wait_step:
+                target = [_to_grid_qubit(q) for q in self.trajectory.qubits]
+                d = freq_map.duration_nanos()
+                if isinstance(d, float):
+                    wait_gate = cirq.WaitGate(
+                        cirq.Duration(nanos=d), qid_shape=cirq.qid_shape(target)
+                    )
+                else:
+                    # The following is patching solution for resolving the parameter
+                    # can be tunits. It should only work for pyle internal translation.
```

Comment:
This seems really hacky and I am worried that it might break assumptions elsewhere, such as in serialization.

Why can't we say something like:
```
elif isinstance(d, tu.Value):
    wait_gate = cirq.WaitGate(cirq.Duration(nanos=d[tu.ns], qid_shape=cirq.qid_shape(target))

Reply:
As discussed offline, it is better to add a new gate that can accept the units